lib/Dialect/TTMetal/Transforms/Passes.cpp

```diff
@@ -16,44 +16,397 @@
+  std::array<int64_t, 2> operator[](CoreCoord coord) const {
+    return data[coord.d * grid[0] * grid[1] + coord.y * grid[1] + coord.x];
+  }
+
+private:
+  std::array<int64_t, 2> grid;
+  SmallVector<std::array<int64_t, 2>> data;
+};
+
+static uint64_t lookupAddress(Value value) {
+  auto blockArg = mlir::dyn_cast<BlockArgument>(value);
+  if (blockArg) {
+    auto funcOp = blockArg.getOwner()->getParentOp();
+    if (mlir::isa<func::FuncOp>(funcOp)) {
+      auto argAlloc = mlir::cast<ArgumentAllocationAttr>(
+          mlir::cast<ArrayAttr>(funcOp->getDiscardableAttr(
+              ArgumentAllocationAttr::name))[blockArg.getArgNumber()]);
+      return argAlloc.getAddress();
+    }
+  }
+
+  auto *op = value.getDefiningOp();
+  if (!op) {
+    return 0;
+  }
+
+  while (isa<DestinationStyleOpInterface>(op)) {
+    assert(op->getResults().size() == 1);
+    auto dps = cast<DestinationStyleOpInterface>(op);
+    assert(dps.getNumDpsInits() == 1);
+    auto *opOperand = dps.getDpsInitOperand(0);
+    value = opOperand->get();
+    op = value.getDefiningOp();
+  }
+
+  auto allocOp = dyn_cast<ttir::AllocOp>(op);
+  if (!allocOp) {
+    return 0;
+  }
+  return allocOp.getAddress();
+}
+
 class TTIRToTTMetalLayoutRewriter : public OpRewritePattern<ttir::ToLayoutOp> {
 public:
   using OpRewritePattern<ttir::ToLayoutOp>::OpRewritePattern;
 
+  struct NocRead {
+    CoreCoord srcCoord;
+    std::int64_t srcOffset = 0;
+    std::int64_t dstOffset = 0;
+    std::int64_t size = 0;
+
+    NocRead(CoreCoord srcCoord, std::int64_t srcOffset, std::int64_t dstOffset,
+            std::int64_t size)
+        : srcCoord(srcCoord), srcOffset(srcOffset), dstOffset(dstOffset),
+          size(size) {}
+
+    bool isContiguous(CoreCoord nextCoord, std::int64_t nextSrcOffset,
+                      std::int64_t nextDstOffset) const {
+      return (nextCoord == srcCoord) && (nextSrcOffset == srcOffset + size) &&
+             (nextDstOffset == dstOffset + size);
+    }
+  };
+
+  llvm::MapVector<CoreCoord, mlir::SmallVector<NocRead>>
+  calculateDataMovement(ArrayRef<std::int64_t> tensorShape,
+                        std::int64_t elemSize, AffineMap src,
+                        AffineMap dst) const {
+    // For now it's just a simple pull model, but eventually we want to leverage
+    // both NoCs and the both read and write
+    llvm::MapVector<CoreCoord, mlir::SmallVector<NocRead>> dst2srcMap;
+    assert(src.getNumResults() == 4);
+    assert(dst.getNumResults() == 4);
+
+    ::ttmlir::utils::sample(tensorShape, [&dst2srcMap, src, dst, elemSize](
+                                             ArrayRef<std::int64_t> index) {
+      auto srcResults = src.compose(index);
+      auto dstResults = dst.compose(index);
+      assert(srcResults.size() == src.getNumResults());
+      assert(dstResults.size() == dst.getNumResults());
+      CoreCoord srcCoord(srcResults);
+      CoreCoord dstCoord(dstResults);
+      std::int64_t srcOffset = srcResults[3] * elemSize;
```

Comment:
Maybe srcResults.back() instead of hardcoded index [3]? There could be more than 4 dims in the mapping in future.